Add IntelRdt to CDI spec

marquiz opened the pull request:
Add support for specifying the OCI Linux.IntelRdt configuration that is the control point for cache and memory bandwidth allocation technologies sucn as Intel RDT (Resource Director Technology). There can only be one IntelRdt configuration per container so in case multiple specs happened to specify it the last one applied prevails. Also, the IntelRdt configuration is always applied as a whole - editing specific sub-fields is not supported.

marquiz commented:
Updated PR:
- Rebased
- Aligned schema fields, now they correspond one-to-one with the OCI spec
- Removed WIP status

zvonkok commented:
@marquiz You may want to enclose CLOS as `CLOS` so that spell check succeeds.

marquiz commented:
> You may want to enclose CLOS as `CLOS` so that spell check succeeds.

Thx, did that. Lovely smart spell checker...

elezar reviewed (changes requested):
Thanks for the work here @marquiz.

I have some questions / comments below.

One thing that is also missing is automatically determining spec version 0.7.0 when calling `MinimumRequiredVersion` in `version.go`.

kad commented on pkg/cdi/container-edits.go:
```diff
@@ -192,6 +204,9 @@ func (e *ContainerEdits) Append(o *ContainerEdits) *ContainerEdits {
 	e.DeviceNodes = append(e.DeviceNodes, o.DeviceNodes...)
 	e.Hooks = append(e.Hooks, o.Hooks...)
 	e.Mounts = append(e.Mounts, o.Mounts...)
+	if o.IntelRdt != nil {
```
intelRDT fiels in OCI spec can't have multiple values, so overwriting already set values theoretically expected behaviour. One thing which is questionable, is to go over individual values, instead of overwriting as a whole struct.

marquiz replied:
Yes, it's like @kad explained. There can only be single RDT configuration per container. Also, the config must be applied as a whole (merging multiple RDT configs together is not something that we want).

elezar replied:
As discussed in the COD working group meeting, I'm happy as long as this is consistent with what is done for other edits (Env, Mounts). It seems to be the case in that the last value applied takes precedence.

We also mentioned that if we do want to expose errors due to conflicts, we can do so in a follow-up. 

One thing that would be good would be to add a unit test to capture this behaviour to prevent regressions.

marquiz commented on SPEC.md:
```diff
@@ -27,7 +27,7 @@ Released versions of the spec are available as Git tags.
 | v0.5.0 |   | Add `HostPath` to `DeviceNodes` |
 | v0.6.0 |   | Add `Annotations` field to `Spec` and `Device` specifications |
-|            |    | Allow dots (`.`)  in name segment of `Kind` field |
+|        |   | Allow dots (`.`)  in name segment of `Kind` field. Add `IntelRdt`field. |
```
Changed as suggested.

marquiz commented on pkg/cdi/container-edits_test.go:
```diff
@@ -245,6 +245,23 @@ func TestValidateContainerEdits(t *testing.T) {
+			name: "invalid rdt config, invalid closID",
+			edits: &cdi.ContainerEdits{
+				IntelRdt: &cdi.IntelRdt{
+					ClosID: "foo/bar",
```
Test cases for these added

marquiz commented on pkg/cdi/container-edits.go:
```diff
@@ -280,6 +295,20 @@ func (m *Mount) Validate() error {
+// ValidateIntelRdt validates the IntelRdt configuration.
+func ValidateIntelRdt(i *specs.IntelRdt) error {
+	// ClosID must be a valid Linux filename
+	if len(i.ClosID) >= 4096 || i.ClosID == "." || i.ClosID == ".." || strings.ContainsAny(i.ClosID, "/\n") {
```
> Is this supposed to check for a path in the current folder?

No, it will practically be a subdir createated under `/sys/fs/resctrl/`. 

> Are there `filepath` functions that can do this for us to make this easier to read?

No, I don't think so. Linux does not have much restrictions on filenames (I think newline is a special case for the `resctrlfs` pseudo-filesystem as is the length). Another possibility is just forget about input validation here and let the low-level runtime engine do it.

marquiz commented:
TODO:
- support version validation
- add unit test(s) for overrides

elezar commented:
@marquiz here is the code to determine the automatic version: https://github.com/cncf-tags/container-device-interface/blob/7b12d28218160cd74dd91f72ad0a355f8dd312e9/pkg/cdi/version.go#L122

elezar commented:
@marquiz I have added the `AdditionalGIDs` extension as discussed yesterday in #179. I would expect some conflicts with your changes, so I'm happy to wait for your PR to land first.

marquiz commented:
Thanks @elezar. I now updated the PR with the TODO items resolved:
1. Support for v0.7.0 version check was added (with unit tests)
2. Unit tests (`TestApply`) extended to cover overriding previous data

In addition I updated the `TestAppend` unit test to cover intelRdt field.

elezar commented on pkg/cdi/container-edits.go:
```diff
@@ -280,6 +295,20 @@ func (m *Mount) Validate() error {
 }
 
+// IntelRdt a wrapper used for validating IntelRdt configuration.
+type IntelRdt struct {
```
nit: Is this type used?

marquiz replied:
Argh, unused indeed, stale type from some early implementation. Well spotted, thanks. Now removed

elezar reviewed (commented):
One last question / comment. 

Otherwise LGTM.

elezar reviewed (approved):
Thanks @marquiz.

LGTM!
